maven-plugin/src/test/java/io/confluent/kafka/schemaregistry/maven/derive/schema/DeriveProtoBufSchemaTest.java

```diff
@@ -0,0 +1,215 @@
+        "    double Double1 = 1;\n" +
+        "    string name = 2;\n" +
+        "  }\n" +
+        "}\n";
+    generateSchemaAndCheckExpected(complexTypesWithPrimitiveValues, expectedSchema);
+  }
+
+  @Test
+  public void testDeriveRecordFailure() {
+    // multiple data types is not allowed in protobuf, should throw error
+    String arrayOfMultipleTypes = "{\"Record\": [1, true]}";
+    assertThrows(IllegalArgumentException.class, () -> generateSchemaAndCheckExpected(arrayOfMultipleTypes, ""));
+
+    // recursive merging of field N with different types should also throw error
+    String recordOfRecords = "{\"RecordOfRecords\": [{\"R1\": {\"N\": 1}}, {\"R1\": {\"N\": \"B\"}}]}";
+    assertThrows(IllegalArgumentException.class, () -> generateSchemaAndCheckExpected(recordOfRecords, ""));
+
+    // nested arrays should raise error
+    String array2D = "{\"Array2d\": [[1, 1.5]]}";
+    assertThrows(IllegalArgumentException.class, () -> generateSchemaAndCheckExpected(array2D, ""));
+  }
+
+  @Test
+  public void testRecursiveMergingOfNumberTypes() throws IOException {
+    // recursive merging of field N with number types should happen
+    String recordOfRecords = "{\"RecordOfNum\": [{\"R1\": {\"N\": 1.5}}, {\"R1\": {\"N\": 1}}]}";
+    generateSchemaAndCheckExpected(recordOfRecords, "syntax = \"proto3\";\n" + "\n" +
+        "message Schema {\n" +
+        "  repeated RecordOfNumMessage RecordOfNum = 1;\n" + "\n" +
+        "  message RecordOfNumMessage {\n" +
+        "    R1Message R1 = 1;\n" + "  \n" +
+        "    message R1Message {\n" +
+        "      double N = 1;\n" +
+        "    }\n" + "  }\n"
+        + "}\n");
+  }
+
+  @Test
+  public void testDeriveMultipleMessages() throws JsonProcessingException {
+    // Message1 and Message2 cannot be merged due to conflicting types of F1
+    // Message3 can merge with both, hence we have 2 different schema generated
+    JsonNode message1 = mapper.readTree("{\"F1\": 1.5, \"F2\": true}");
+    JsonNode message2 = mapper.readTree("{\"F1\": 1, \"F2\": 1}");
+    JsonNode message3 = mapper.readTree("{\"F3\": [1, 1.5, 3]}");
+    JsonNode schema = derive.getSchemaForMultipleMessages(Arrays.asList(message1, message2, message3)).get("schemas");
+
+    String expectedSchema1 = "syntax = \"proto3\";\n\n" +
+        "message Schema {\n" +
+        "  double F1 = 1;\n" +
+        "  bool F2 = 2;\n" +
+        "  repeated double F3 = 3;\n" +
+        "}\n";
+    assertEquals(schema.get(0).get("schema").asText(), expectedSchema1);
+    assertEquals(schema.get(0).get("messagesMatched").toString(), "[0,2]");
+
+    String expectedSchema2 = "syntax = \"proto3\";\n\n" +
+        "message Schema {\n" +
+        "  double F1 = 1;\n" +
+        "  int32 F2 = 2;\n" +
+        "  repeated double F3 = 3;\n" +
+        "}\n";
+    assertEquals(schema.get(1).get("schema").asText(), expectedSchema2);
+    assertEquals(schema.get(1).get("messagesMatched").toString(), "[1,2]");
+  }
```

Comment:
is test for mergeArrays function not required for a DeriveProtobufSchema object. I think we need it because there is an implementation for mergeMultipleDataTypes function?

Reply:
I've added a couple of tests